core/src/main/scala/kafka/log/Log.scala

```diff
@@ -465,6 +473,9 @@ class Log(@volatile var dir: File,
       // check that offsets are monotonically increasing
       if(lastOffset >= entry.offset)
         monotonic = false
+      // keep track of the highest offset seen
+      if (maxOffset < entry.offset)
+        maxOffset = entry.offset
```

Comment:
Yes, offsets are guaranteed to be monotonically increasing. If not, we throw an exception in Log.append. Timestamp, on the other hand, may not be monotonically increasing.